src/amr/solvers/solver_ppc.hpp

```diff
@@ -218,373 +224,249 @@ void SolverPPC<HybridModel, AMR_Types>::restoreState_(level_t& level, Ions& ions
-            electrons.update(layout);
-            auto& Ve = electrons.velocity();
-            auto& Ne = electrons.density();
-            auto& Pe = electrons.pressure();
-            auto __  = core::SetLayout(&layout, ohm_);
-            ohm_(Ne, Ve, Pe, Bpred, J, Epred);
-            resourcesManager->setTime(Epred, *patch, newTime);
-        }
+        for (auto& state : view)
+            state.electrons.update(state.layout);
+        ohm_(view.layouts, view.electrons_N, view.electrons_Ve, view.electrons_Pe,
+             view.electromagPred_B, view.J, view.electromagPred_E);
+        for (auto& state : view)
+            resourcesManager->setTime(state.electromagPred.E, *state.patch, newTime);
     }
 }
 
 
 template<typename HybridModel, typename AMR_Types>
-void SolverPPC<HybridModel, AMR_Types>::predictor2_(level_t& level, HybridModel& model,
+void SolverPPC<HybridModel, AMR_Types>::predictor2_(level_t& level, HybridModelView_t& view,
                                                     Messenger& fromCoarser,
                                                     double const currentTime, double const newTime)
 {
     PHARE_LOG_SCOPE("SolverPPC::predictor2_");
 
-    auto& hybridState      = model.state;
-    auto& resourcesManager = model.resourcesManager;
     auto dt                = newTime - currentTime;
-    auto levelNumber       = level.getLevelNumber();
-
-
+    auto& resourcesManager = view.model().resourcesManager;
 
     {
         PHARE_LOG_SCOPE("SolverPPC::predictor2_.faraday");
+        faraday_(view.layouts, view.electromag_B, view.electromagAvg_E, view.electromagPred_B, dt);
 
-        auto& Bpred = electromagPred_.B;
-        auto& B     = hybridState.electromag.B;
-        auto& Eavg  = electromagAvg_.E;
-
-        for (auto& patch : level)
-        {
-            auto _      = resourcesManager->setOnPatch(*patch, Bpred, B, Eavg);
-            auto layout = PHARE::amr::layoutFromPatch<GridLayout>(*patch);
-            auto __     = core::SetLayout(&layout, faraday_);
-            faraday_(B, Eavg, Bpred, dt);
-
-            resourcesManager->setTime(Bpred, *patch, newTime);
-        }
+        for (auto& state : view)
+            resourcesManager->setTime(state.electromagPred.B, *state.patch, newTime);
     }
 
-
     {
         PHARE_LOG_SCOPE("SolverPPC::predictor2_.ampere");
-
-        auto& Bpred = electromagPred_.B;
-        auto& J     = hybridState.J;
-
-
-        for (auto& patch : level)
-        {
-            auto _      = resourcesManager->setOnPatch(*patch, Bpred, J);
-            auto layout = PHARE::amr::layoutFromPatch<GridLayout>(*patch);
-            auto __     = core::SetLayout(&layout, ampere_);
-            ampere_(Bpred, J);
-
-            resourcesManager->setTime(J, *patch, newTime);
-        }
-        fromCoarser.fillCurrentGhosts(J, levelNumber, newTime);
+        ampere_(view.layouts, view.electromagPred_B, view.J);
+        for (auto& state : view)
```

Comment:
not sure what form you specifically have in mind, but there's probably a way to hide teh loop sure

Reply:
nothing very specific maybe just some free function to avoid repeating the loop, we've made the effort to remove the other one and move it into the transformer, maybe we can hide the setTime one too while we're at it. Just saying..